lib/bolt/catalog.rb

```diff
@@ -71,8 +71,11 @@ def compile_catalog(request)
             Puppet.lookup(:pal_current_node).trusted_data = target['trusted']
             pal.with_catalog_compiler do |compiler|
               # This needs to happen inside the catalog compiler so loaders are initialized for loading
-              vars = Puppet::Pops::Serialization::FromDataConverter.convert(request['plan_vars'])
-              pal.send(:add_variables, compiler.send(:topscope), target['variables'].merge(vars))
+              plan_vars = Puppet::Pops::Serialization::FromDataConverter.convert(request['plan_vars'])
+              merged = target['variables'].merge(plan_vars)
+              merged.reject! { |k, _| target['facts'].keys.include?(k) }
```

Comment:
A comment explaining this would be helpful. If I understand correctly, we exclude any variables that share the name of a fact, and then later Puppet will automatically assign variables for each fact. Is that correct or is the order reversed (Puppet has already assigned variables for the facts and we need to avoid assigning them here)?

Reply:
The original issue also mentions that we want to warn here that a variable in the current scope will be shadowed by a target fact of the same name.

Reply:
M'k, this should be logged and commented now